### PR TITLE
platform/jimfuqian/10488 - Add functions listing blobs used by the pipeline

### PR DESCRIPTION
**Pull request by JFU-NAVA-PBC**
This PR:

Added functions to BlobAccess to list blobs used by universal pipeline.

Test Steps:
1. *Pass all existing tests*

## Changes
- *Added below functions in BlobAccess:*
- "BlobAccess.listBlobsReactive(directory, blobConnInfo):  PagedIterable<BlobItem>"
- "BlobAccess.listBlobs(directory, blobConnInfo): List<BlobEntry>"
- *Added below enum to facilitate forming UP steps related blob container 'folder's*
- "enum class UniversalpipelineSteps(val step: String) - see Files changed for details"
- *(For web UI changes) Include screenshots/video of changes*

### Testing
- [x] Tested locally?
- [x] Ran `./prime test` or `./gradlew testSmoke` against local Docker ReportStream container?
- [ ] (For Changes to /frontend-react/...) Ran `npm run lint:write`? 
- [X] Added tests? Added unit test

### Process
- [ ] Are there licensing issues with any new dependencies introduced?
- [ ] Includes a summary of what a code reviewer should test/verify?
- [ ] Updated the release notes?
- [ ] Database changes are submitted as a separate PR?
- [ ] DevOps team has been notified if PR requires ops support?

## Linked Issues
- #10488

## To Be Done
*Create GitHub issues to track the work remaining, if any*
- #issue

## Specific Security-related subjects a reviewer should pay specific attention to
- Does this PR introduce new endpoints?
    - new endpoint A
    - new endpoint B
- Does this PR include changes in authentication and/or authorization of existing endpoints?
- Does this change introduce new dependencies that need vetting?
- Does this change require changes to our infrastructure?
- Does logging contain sensitive data?
- Does this PR include or remove any sensitive information itself?

If you answered '_yes_' to any of the questions above, conduct a detailed Review that addresses at least:

- What are the potential security threats and mitigations? Please list the _STRIDE_ threats and how they are mitigated
    - **S**poofing (faking authenticity)
        - Threat _T_, which could be achieved by _A_, is mitigated by _M_
    - **T**ampering (influence or sabotage the integrity of information, data, or system)
    - **R**epudiation (the ability to dispute the origin or originator of an action)
    - **I**nformation disclosure (data made available to entities who should not have it)
    - **D**enial of service (make a resource unavailable)
    - **E**levation of Privilege (reduce restrictions that apply or gain privileges one should not have)
- Have you ensured logging does not contain sensitive data?
- Have you received any additional approvals needed for this change?

### CODE REVIEW

File: prime-router/src/main/kotlin/azure/BlobAccess.kt [hunk 1]
```diff
@@ -266,7 +328,7 @@ class BlobAccess() : Logging {
          * If one exists for the container name and connection string, the existing one will be reused.
          * @return the blob container client
          */
-        private fun getBlobContainer(blobConnInfo: BlobContainerMetadata): BlobContainerClient {
+        fun getBlobContainer(blobConnInfo: BlobContainerMetadata): BlobContainerClient {
```
Comment (jack-h-wang) [inline prime-router/src/main/kotlin/azure/BlobAccess.kt:331]: This should probably stay private

```suggestion
        private fun getBlobContainer(blobConnInfo: BlobContainerMetadata): BlobContainerClient {
```

File: prime-router/src/main/kotlin/azure/BlobAccess.kt [hunk 2]
```diff
@@ -188,6 +193,63 @@ class BlobAccess() : Logging {
             return getBlobClient(blobUrl, blobConnInfo).exists()
         }
 
+        /**
+         * Data class for returning the current version of a blob
+         * and optionally it's past versions
```
Comment (jack-h-wang) [inline prime-router/src/main/kotlin/azure/BlobAccess.kt:198]:
```suggestion
         * and optionally its past versions
```

File: prime-router/src/main/kotlin/azure/BlobAccess.kt [hunk 3]
```diff
@@ -188,6 +193,63 @@ class BlobAccess() : Logging {
...
+         * If the following blobs were stored
+         * foo/item.txt
+         * foo/baz/item2.txt
+         * bar/item3.txt
+         *
+         * using "foo" as the directory would return foo/item.txt and foo/baz/item2.txt
+         *
+         * @param directory - the prefix to filter the blobs by
+         * @param blobConnInfo - metadata of which azure storage account to read from
+         * @param includeVersion - whether or not to return the past versions of the blobs
```
Comment (jack-h-wang) [inline prime-router/src/main/kotlin/azure/BlobAccess.kt:218]:
```suggestion
         * @param includeVersion - whether to return the past versions of the blobs
```

**Review by jack-h-wang — APPROVED** [review]
Some minor nits but makes sense to me and all tests pass

**Review by arnejduranovic — APPROVED** [review]

File: prime-router/src/main/kotlin/azure/BlobAccess.kt [hunk 4]
```diff
@@ -19,9 +22,11 @@ import java.net.URL
 import java.net.URLDecoder
 import java.nio.charset.Charset
 import java.security.MessageDigest
+import java.time.Duration
 
 const val defaultBlobContainerName = "reports"
 const val defaultBlobDownloadRetryCount = 5
+const val listBlobTimeoutSeconds: Long = 30
```
Comment (jalbinson) [inline prime-router/src/main/kotlin/azure/BlobAccess.kt:29]: If this operation could actually take 30 seconds I think we should consider making it async.

**Pull request merged**